pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -307,30 +316,54 @@ func (e *endpointSliceCache) Update(hostname host.Name, slice string, endpoints
 	e.mu.Lock()
 	defer e.mu.Unlock()
 	if len(endpoints) == 0 {
-		delete(e.endpointsByServiceAndSlice[hostname], slice)
+		for _, ip := range e.endpointKeysByServiceAndSlice[hostname][slice] {
+			delete(e.endpointByKey, ip)
+		}
+		delete(e.endpointKeysByServiceAndSlice[hostname], slice)
+	}
+	if _, f := e.endpointKeysByServiceAndSlice[hostname]; !f {
+		e.endpointKeysByServiceAndSlice[hostname] = make(map[string][]endpointKey)
 	}
-	if _, f := e.endpointsByServiceAndSlice[hostname]; !f {
-		e.endpointsByServiceAndSlice[hostname] = make(map[string][]*model.IstioEndpoint)
+	keys := make([]endpointKey, 0, len(endpoints))
+	for _, ep := range endpoints {
+		key := endpointKey{ep.Address, ep.ServicePortName}
+		keys = append(keys, key)
+		// We will always overwrite. A conflict here means an endpoint is transitioning
+		// from one slice to another See
+		// https://github.com/kubernetes/website/blob/master/content/en/docs/concepts/services-networking/endpoint-slices.md#duplicate-endpoints
+		// In this case, we can always assume and update is fresh, although older slices
```

Comment:
s/"we can always assume and update is fresh"/ "we can always assume that update is fresh"